lncli: group commands into categories

guggero opened the pull request:
The number of commands in the `lncli` utility continually grows and it is getting harder and harder to look through the (unsorted) list to find the command one is looking for.

Fortunately, the command line library that is used supports grouping commands into categories.
That's what I did. Here is how the output looks now:

```text
NAME:
   lncli - control plane for your Lightning Network Daemon (lnd)

USAGE:
   lncli [global options] command [command options] [arguments...]
   
VERSION:
   0.4.1 commit=a6341d7e1ee9433cc653e3cc77eff57c054f6fe3
   
COMMANDS:
     getinfo     Returns basic information related to the active daemon
     debuglevel  Set the debug level.
     stop        Stop and shutdown the daemon.
     help, h     Shows a list of commands or help for one command

   Channels:
     openchannel       Open a channel to a node or an existing peer.
     closechannel      Close an existing channel.
     closeallchannels  Close all existing channels.
     channelbalance    Returns the sum of the total available channel balance across all open channels
     pendingchannels   Display information pertaining to pending channels
     listchannels      List all open channels
     getchaninfo       Get the state of a channel
     getnetworkinfo    Getnetworkinfo
     feereport         Display the current fee policies of all active channels
     updatechanpolicy  Update the channel policy for all channels, or a single channel

   On-chain:
     sendmany       Send bitcoin on-chain to multiple addresses.
     sendcoins      Send bitcoin on-chain to an address
     listchaintxns  List transactions from the wallet.

   Payments:
     sendpayment    Send a payment over lightning
     payinvoice     Pay an invoice over lightning
     addinvoice     Add a new invoice.
     lookupinvoice  Lookup an existing invoice by its payment hash.
     listinvoices   List all invoices currently stored.
     listpayments   List all outgoing payments
     queryroutes    Query a route to a destination.
     decodepayreq   Decode a payment request.
     fwdinghistory  Query the history of all forwarded htlcs

   Peers:
     connect        Connect to a remote lnd peer
     disconnect     Disconnect a remote lnd peer identified by public key
     listpeers      List all active, currently connected peers.
     describegraph  Describe the network graph
     getnodeinfo    Get information on a specific node.

   Wallet:
     create         
     unlock         
     newaddress     Generates a new address.
     walletbalance  Compute and display the wallet's current balance
     signmessage    Sign a message with the node's private key
     verifymessage  Verify a message signed with the signature

GLOBAL OPTIONS:
   --rpcserver value        host:port of ln daemon (default: "localhost:10009")
   --lnddir value           path to lnd's base directory (default: "/home/guggero/.lnd")
   --tlscertpath value      path to TLS certificate (default: "/home/guggero/.lnd/tls.cert")
   --no-macaroons           disable macaroon authentication
   --macaroonpath value     path to macaroon file (default: "/home/guggero/.lnd/admin.macaroon")
   --macaroontimeout value  anti-replay macaroon validity time in seconds (default: 60)
   --macaroonip value       if set, lock macaroon to specific IP address
   --help, -h               show help
   --version, -v            print the version
   
```

Feel free to suggest different category names or different groupings. But I think it's long overdue to add categories.

halseth commented:
OOOooOh, this is pure beauty! One suggestion I have is to move all `WalletUnlocker` commands into their own category (`unlock`, `create`, and soon `changepassword`). These commands can only be used at `lnd` startup, is is possible to add that as a comment to the category?

guggero commented:
I'm glad you like it!

I've grouped the commands you mentioned in a group called `Startup` and added a usage comment to them.
Then I also made sure that every usage comment is a full sentence to get a more uniform look.

I updated the full output in my initial text to not clutter the comment timeline.

05nelsonm commented:
kind of a noob question, but how might one interact with these cli commands using a bash script? Trying to write a script for a graceful shutdown and need to execute `lncli stop`. Basic example looks like:

```
#!/bin/bash
ssh -p XXXX user@ip-address 'lncli stop'
```

but keep getting `bash: lncli: command not found` returned to me. Any suggestions?

guggero commented:
The error message you‘re getting is from the SSH/Bash session on the remote computer. It can‘t find your `lncli` binary in the `PATH` environment variable.
Make sure you add the directory where your LND binaries reside to PATH.
You can do so in `~/.bashrc` for example.

> On 6 May 2019, at 18:17, 05nelsonm <notifications@github.com> wrote:
> 
> kind of a noob question, but how might one interact with these cli commands using a bash script? Trying to write a script for a graceful shutdown and need to execute lncli stop. Basic example looks like:
> 
> #!/bin/bash
> ssh -p XXXX user@ip-address 'lncli stop'
> but keep getting bash: lncli: command not found returned to me. Any suggestions?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

05nelsonm commented:
I'm still not sure on how to get the scripts to work as I've already modified my `~/.bashrc` file to include:

```
export GOPATH=~/gocode
export PATH=$PATH:$GOPATH/bin
```

I can execute `lncli whatever` commands from gnome-terminal, but when executing from a script (remotely, or even on the machine) it just returns `bash: lncli: command not found`. Is there an additional `PATH` that I need to express in `~/.bashrc`, or change the header of the script from `#!/bin/bash` to something else such that the commands are executed using environment variables located in `~/.bashrc`?

Sorry for the nub q's, I'm trying to learn and appreciate your time greatly! Thanks.

> The error message you‘re getting is from the SSH/Bash session on the remote computer. It can‘t find your `lncli` binary in the `PATH` environment variable. Make sure you add the directory where your LND binaries reside to PATH. You can do so in `~/.bashrc` for example.
> […](#)
> On 6 May 2019, at 18:17, 05nelsonm ***@***.***> wrote: kind of a noob question, but how might one interact with these cli commands using a bash script? Trying to write a script for a graceful shutdown and need to execute lncli stop. Basic example looks like: #!/bin/bash ssh -p XXXX ***@***.*** 'lncli stop' but keep getting bash: lncli: command not found returned to me. Any suggestions? — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or mute the thread.

05nelsonm commented:
figured it out...

```
#!/bin/bash

ssh -p XXXX user@ip-address '/home/user/gocode/bin/lncli getnetworkinfo'
```

thx!

guggero commented:
I'm sorry, what I wrote before was incorrect. I had to research this too. Aparently remote sessions with SSH are non-interactive. And the `.bashrc` script exits in the first few lines if the session is non-interactive.

For more information have a look here: https://stackoverflow.com/questions/940533/how-do-i-set-path-such-that-ssh-userhost-command-works

05nelsonm commented:
Came back to it this evening and figured out what was up. 

Installation instructions for golang (https://golang.org/doc/install) say to modify `/etc/profile` by adding:
`export PATH=$PATH:/usr/local/go/bin`

Guess I missed that.......

Thanks for your help @guggero
